CI: kernel: Improve kernel github action

hauke opened the pull request:
This fixes multiple problems we saw when using the kernel github action. 
The Action is pretty unstable and I hope this will improve it: https://github.com/openwrt/openwrt/actions/workflows/kernel.yml

hauke committed:
CI: kernel: Trigger workflow for more directories

Trigger the kernel build workflow also for more directories.

Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

hauke committed:
CI: kernel: Use downloads.cdn.openwrt.org

Use downloads.cdn.openwrt.org to download the toolchain. This should
reduce the load on the servers.

Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

hauke committed:
CI: kernel: Show used OpenWrt configuration

Show the configuration used to build OpenWrt before starting the build.
This should make it easier for people to reproduce problems when it
fails.

Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

aparcar commented on .github/workflows/kernel.yml:
```diff
@@ -68,13 +68,6 @@ jobs:
           echo "TARGET=$TARGET" >> "$GITHUB_ENV"
           echo "SUBTARGET=$SUBTARGET" >> "$GITHUB_ENV"
 
-      - name: Update & Install feeds
```
There are Kernel modules in the feeds which are no longer tested once this is removed. I'm fine removing this but then again we should check if changed Kernels break other peoples packages.

Another option could be to use GitHubs mirrors of our feeds instead of git.openwrt.org

hauke replied:
I think we are only building the kernel modules which are activated by default and not all of them . None of the kernel modules from the feed should be active by default. 
You have to add this to the configuration to build all kernel modules: `CONFIG_ALL_KMODS=y`

aparcar replied:
We set the configuration option on our buildbots

hauke replied:
No it does not use the configuration the build bots are using.
It uses this configuration for example:
````
CONFIG_TARGET_apm821xx=y
CONFIG_TARGET_apm821xx_nand=y
CONFIG_TARGET_apm821xx_nand_Default=y
CONFIG_DEVEL=y
CONFIG_BUILD_NLS=y
CONFIG_EXTERNAL_TOOLCHAIN=y
CONFIG_EXTERNAL_TOOLCHAIN_LIBC_USE_MUSL=y
CONFIG_LIBC_FILE_SPEC="./powerpc-openwrt-linux-musl/lib/ld-musl-powerpc.so.1 ./powerpc-openwrt-linux-musl/lib/libc.so"
CONFIG_LIBC_ROOT_DIR="/__w/openwrt/openwrt/openwrt/openwrt-toolchain-apm821xx-nand_gcc-11.3.0_musl.Linux-x86_64/toolchain-powerpc_464fp_gcc-11.3.0_musl"
CONFIG_LIBGCC_FILE_SPEC="./powerpc-openwrt-linux-musl/lib/libgcc_s.so ./powerpc-openwrt-linux-musl/lib/libgcc_s.so.1"
CONFIG_LIBGCC_ROOT_DIR="/__w/openwrt/openwrt/openwrt/openwrt-toolchain-apm821xx-nand_gcc-11.3.0_musl.Linux-x86_64/toolchain-powerpc_464fp_gcc-11.3.0_musl"
CONFIG_LIBPTHREAD_FILE_SPEC="./lib/libpthread{-*.so,.so.*}"
CONFIG_LIBPTHREAD_ROOT_DIR="/__w/openwrt/openwrt/openwrt/openwrt-toolchain-apm821xx-nand_gcc-11.3.0_musl.Linux-x86_64/toolchain-powerpc_464fp_gcc-11.3.0_musl"
CONFIG_LIBRT_FILE_SPEC="./lib/librt{-*.so,.so.*}"
CONFIG_LIBRT_ROOT_DIR="/__w/openwrt/openwrt/openwrt/openwrt-toolchain-apm821xx-nand_gcc-11.3.0_musl.Linux-x86_64/toolchain-powerpc_464fp_gcc-11.3.0_musl"
CONFIG_LIBSTDCPP_FILE_SPEC="./powerpc-openwrt-linux-musl/lib/libstdc++.so ./powerpc-openwrt-linux-musl/lib/libstdc++.so.6 ./powerpc-openwrt-linux-musl/lib/libstdc++.so.6.0.29 ./powerpc-openwrt-linux-musl/lib/libstdc++.so.6.0.29-gdb.py"
CONFIG_LIBSTDCPP_ROOT_DIR="/__w/openwrt/openwrt/openwrt/openwrt-toolchain-apm821xx-nand_gcc-11.3.0_musl.Linux-x86_64/toolchain-powerpc_464fp_gcc-11.3.0_musl"
CONFIG_PACKAGE_libstdcpp=y
CONFIG_STRIP_ARGS="--strip-all"
CONFIG_TARGET_NAME="powerpc-openwrt-linux-musl"
CONFIG_TOOLCHAIN_BIN_PATH="./usr/bin ./bin"
CONFIG_TOOLCHAIN_INC_PATH="./usr/include ./include"
CONFIG_TOOLCHAIN_LIBC="musl"
CONFIG_TOOLCHAIN_LIB_PATH="./usr/lib ./lib"
CONFIG_TOOLCHAIN_PREFIX="powerpc-openwrt-linux-musl-"
CONFIG_TOOLCHAIN_ROOT="/__w/openwrt/openwrt/openwrt/openwrt-toolchain-apm821xx-nand_gcc-11.3.0_musl.Linux-x86_64/toolchain-powerpc_464fp_gcc-11.3.0_musl"
CONFIG_USE_EXTERNAL_LIBC=y
# CONFIG_USE_SSTRIP is not set
CONFIG_USE_STRIP=y
````

The build bots are using this configuration: https://downloads.openwrt.org/snapshots/targets/apm821xx/nand/config.buildinfo

aparcar replied:
Sorry my point was that the build bots are using the all kmods option and so should the CI. I think using GitHub repo mirrors fixes the issue in a better way than ignoring those external kmods

hauke replied:
I will try to remove the feed URLs using sed in the feeds.conf instead of deleting it.
I think we do not use any package from the feeds in these builds.

hauke commented:
One of the builds failed like this:
````
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether UID '1000' is supported by ustar format... yes
checking whether GID '1000' is supported by ustar format... yes
checking how to create a ustar tar archive... gnutar
checking for x86_64-pc-linux-gnu-gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of gcc... gcc3
checking for x86_64-pc-linux-gnu-ranlib... no
checking for ranlib... ranlib
checking build system type... x86_64-pc-linux-gnu
checking host system type... build-aux/config.sub: line 54: sage:: command not found
build-aux/config.sub: line 56: Canonicalize: command not found
build-aux/config.sub: line 58: Options:: command not found
build-aux/config.sub: line 59: -h,: command not found
build-aux/config.sub: line 60: -t,: command not found
build-aux/config.sub: line 61: -v,: command not found
build-aux/config.sub: line 66: syntax error near unexpected token `('
build-aux/config.sub: line 66: `GNU config.sub ($timestamp)'
configure: error: /usr/bin/env bash build-aux/config.sub x86_64-pc-linux-gnu failed
make[2]: *** [Makefile:31: /__w/openwrt/openwrt/openwrt/build_dir/host/gengetopt-2.23/.configured] Error 2
time: tools/gengetopt/compile#11.34#0.65#13.39
````
I do not understand this error.

aparcar commented:
@hauke We do have some kmods in the packages.git feed, right?

```
❯ rg KernelPackage
kernel/ovpn-dco/Makefile
26:define KernelPackage/ovpn-dco
34:define KernelPackage/ovpn-dco/description
61:$(eval $(call KernelPackage,ovpn-dco))

kernel/mdio-netlink/Makefile
19:define KernelPackage/mdio-netlink
30:define KernelPackage/mdio-netlink/description
38:$(eval $(call KernelPackage,mdio-netlink))

kernel/macremapper/Makefile
25:define KernelPackage/macremapper
35:define KernelPackage/macremapper/description
42:$(eval $(call KernelPackage,macremapper))

kernel/mtd-rw/Makefile
28:define KernelPackage/mtd-rw
35:define KernelPackage/mtd-rw/description
51:$(eval $(call KernelPackage,mtd-rw))

kernel/v4l2loopback/Makefile
21:define KernelPackage/v4l2loopback
29:define KernelPackage/v4l2loopback/description
49:$(eval $(call KernelPackage,v4l2loopback))

kernel/antfs/Makefile
18:define KernelPackage/fs-antfs
26:define KernelPackage/fs-antfs/description
44:$(eval $(call KernelPackage,fs-antfs))

utils/gl-mifi-mcu/Makefile
13:define KernelPackage/gl-mifi-mcu
20:define KernelPackage/gl-mifi-mcu/description
36:$(eval $(call KernelPackage,gl-mifi-mcu))

libs/xr_usb_serial_common/Makefile
20:define KernelPackage/usb-serial-xr_usb_serial_common
39:$(eval $(call KernelPackage,usb-serial-xr_usb_serial_common))

libs/libpfring/Makefile
48:define KernelPackage/pf-ring
56:define KernelPackage/pf-ring/description
90:$(eval $(call KernelPackage,pf-ring))

libs/dmx_usb_module/Makefile
24:define KernelPackage/usb-serial-dmx_usb_module
35:define KernelPackage/usb-serial-dmx_usb_module/description
55:$(eval $(call KernelPackage,usb-serial-dmx_usb_module))

admin/netatop/Makefile
20:define KernelPackage/netatop
30:define KernelPackage/netatop/description
85:$(eval $(call KernelPackage,netatop))

net/openvswitch/Makefile
59:  define KernelPackage/$(call ovs_kmod_package_name,$(1))
278:  $(eval $(call KernelPackage,$(p)))\

net/xtables-addons/Makefile
99:  define KernelPackage/ipt-$(1)
107:  $$(eval $$(call KernelPackage,ipt-$(1)))

net/siit/Makefile
18:define KernelPackage/siit
29:define KernelPackage/siit/description
42:$(eval $(call KernelPackage,siit))

net/jool/Makefile
78:define KernelPackage/jool-netfilter
95:define KernelPackage/jool-netfilter/description
144:$(eval $(call KernelPackage,jool-netfilter))

net/coova-chilli/Makefile
68:define KernelPackage/ipt-coova
77:define KernelPackage/ipt-coova/description
148:$(eval $(call KernelPackage,ipt-coova))

devel/lttng-modules/Makefile
25:define KernelPackage/lttng
45:$(eval $(call KernelPackage,lttng))
```

hauke commented:
Yes there are some kernel packages in the feed. Do you want to check that all kernel modules are building in this workflow?

hauke committed:
CI: kernel: Checkout feeds from github

Instead of cloning the feeds from the default location at
git.openwrt.org use the github action to clone them directly from
github. We saw some error messages when cloning from git.openwrt.org,
probably related to some rate limiting applied. Cloning from github
within a github action should work more stable.

The "./scripts/feeds update -a" script will use the already checked out
feed repositories and not clone them again from git.openwrt.org, but it
will also not change the branch name.

Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

hauke commented:
I changed the code to clone from github using the `actions/checkout@v2` action, lets see if it works. 
If you want to build all kernel modules you have to add this: `echo CONFIG_ALL_KMODS=y >> .config`

aparcar commented:
I'd suggest to add CONFIG_ALL_KMODS, if not we end up not testing everything we offer...

Biggest chunk is currently the `tools/compile` step, I'll see to improve that. Instead of building in the regular builedbot container, we could create a build boot-tool container ontop of that.

hauke committed:
CI: kernel: Build all kernel modules

Activate building all kernel modules.

This builds all kernel modules from the core packages and the feeds.

Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

hauke commented:
Thank you for the patches, I applied them to master in 546822775cd988360964687c0cb430d67bd5c617, 7b4daf00766620faac8212b46259064ca6e2c716, 43afaf6149129a72a8f3e5d3d0d545e54ac81b3a, b120e78917099d46a25cc521998b917fdf08e388 and fed325f4633cb302a5f43c42285e8ab0fd144262
